x/cron/keeper/keeper.go

```diff
@@ -0,0 +1,218 @@
+// and executes messages in each one
+// NOTE that errors in contract calls DO NOT stop schedule execution
+func (k *Keeper) ExecuteReadySchedules(ctx sdk.Context) {
+	telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), LabelCheckTimer)
+	schedules := k.getSchedulesReadyForExecution(ctx)
+
+	for _, schedule := range schedules {
+		k.executeSchedule(ctx, schedule)
+	}
+}
+
+// AddSchedule adds new schedule to execution for every block `period`.
+// First schedule execution is supposed to be on `now + period` block.
+func (k *Keeper) AddSchedule(ctx sdk.Context, name string, period uint64, msgs []types.MsgExecuteContract) error {
+	if k.scheduleExists(ctx, name) {
+		return fmt.Errorf("schedule already exists with name=%v", name)
+	}
+
+	schedule := types.Schedule{
+		Name:              name,
+		Period:            period,
+		Msgs:              msgs,
+		LastExecuteHeight: uint64(ctx.BlockHeight()), // let's execute newly added schedule on `now + period` block
+	}
+	k.storeSchedule(ctx, schedule)
+
+	return nil
+}
+
+// RemoveSchedule removes schedule with a given `name`
+func (k *Keeper) RemoveSchedule(ctx sdk.Context, name string) {
+	k.removeSchedule(ctx, name)
+}
+
+// GetSchedule returns schedule with a given `name`
+func (k *Keeper) GetSchedule(ctx sdk.Context, name string) (*types.Schedule, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+	bzSchedule := store.Get(types.GetScheduleKey(name))
+	if bzSchedule == nil {
+		return nil, false
+	}
+
+	var schedule types.Schedule
+	k.cdc.MustUnmarshal(bzSchedule, &schedule)
+	return &schedule, true
+}
+
+// GetAllSchedules returns all schedules
+func (k *Keeper) GetAllSchedules(ctx sdk.Context) []types.Schedule {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+
+	res := make([]types.Schedule, 0)
+
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var schedule types.Schedule
+		k.cdc.MustUnmarshal(iterator.Value(), &schedule)
+		res = append(res, schedule)
+	}
+
+	return res
+}
+
+func (k *Keeper) getSchedulesReadyForExecution(ctx sdk.Context) []types.Schedule {
+	params := k.GetParams(ctx)
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.ScheduleKey)
+	count := uint64(0)
+
+	res := make([]types.Schedule, 0)
+
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var schedule types.Schedule
+		k.cdc.MustUnmarshal(iterator.Value(), &schedule)
+
+		if k.intervalPassed(ctx, schedule) {
+			res = append(res, schedule)
+			count++
+
+			if count >= params.Limit {
+				k.Logger(ctx).Info("limit of schedule executions per block reached")
+				return res
+			}
+		}
+	}
+
+	return res
+}
+
+// executeSchedule executes given schedule and changes LastExecuteHeight
+func (k *Keeper) executeSchedule(ctx sdk.Context, schedule types.Schedule) {
+	for idx, msg := range schedule.Msgs {
```

Comment:
nope, this implementation does not work, unfortunately. if we allow for execution of multiple messages, we need to guarantee transactionality. please use a cached context to execute messages from a schedule one by one, and only update the state if all of the messages were executed successfully. in case of any questions, talk to @pr0n00gler